sycl/test-e2e/Basic/vector/int-convert.cpp

```diff
@@ -0,0 +1,148 @@
+  if (!all_good) {
+    DEBUG_PRINT("host and device results mismatch:");
+    DEBUG_PRINT(
+        "\t{" << static_cast<int>(a.x()) << ", " << static_cast<int>(a.y())
+              << ", " << static_cast<int>(a.z()) << ", "
+              << static_cast<int>(a.w()) << "} vs {" << static_cast<int>(b.x())
+              << ", " << static_cast<int>(b.y()) << ", "
+              << static_cast<int>(b.z()) << ", " << static_cast<int>(b.w())
+              << "}");
+  }
+  assert(all_good);
+}
+
+template <typename From, typename To> void check_convert() {
+  DEBUG_PRINT("checking vec<" << to_string<From>() << ", 4>::convert<"
+                              << to_string<To>() << ">()");
+
+  sycl::vec<From, 4> input;
+  if constexpr (std::is_signed_v<From>) {
+    input = sycl::vec<From, 4>{static_cast<From>(37), static_cast<From>(0),
+                               static_cast<From>(-11), static_cast<From>(13)};
+  } else {
+    input = sycl::vec<From, 4>{static_cast<From>(37), static_cast<From>(0),
+                               static_cast<From>(11), static_cast<From>(13)};
+  }
+
+  sycl::vec<To, 4> hostResult = input.template convert<To>();
+
+  sycl::buffer<sycl::vec<To, 4>> buf(sycl::range{1});
+  sycl::queue q;
+  q.submit([&](sycl::handler &cgh) {
+     sycl::accessor acc(buf, cgh);
+     cgh.single_task([=]() { acc[0] = input.template convert<To>(); });
+   }).wait();
+
+  auto acc = buf.get_host_access();
+  auto deviceResult = acc[0];
+
+  // Host and device results must match.
+  check_vectors_equal(hostResult, deviceResult);
+
+  // And they should match with a reference, which is for integer conversions
+  // can be computed with a simple static_cast.
+  // Strictly speaking, integer conversions are underspecified in the SYCL 2020
+  // spec, but `static_cast` implementation matches SYCL-CTS, so we will leave
+  // it here for now as well.
+  // See https://github.com/KhronosGroup/SYCL-Docs/issues/492
+  assert(deviceResult.x() == static_cast<To>(input.x()));
+  assert(deviceResult.y() == static_cast<To>(input.y()));
+  assert(deviceResult.z() == static_cast<To>(input.z()));
+  assert(deviceResult.w() == static_cast<To>(input.w()));
+}
+
+template <class T>
+constexpr auto has_unsigned_v =
+    std::is_integral_v<T> && !std::is_same_v<T, bool> &&
+    !std::is_same_v<T, sycl::byte> && !std::is_same_v<T, std::byte>;
+
+template <typename From, typename To> void check_signed_unsigned_convert_to() {
+  check_convert<From, To>();
+  if constexpr (has_unsigned_v<To>)
+    check_convert<From, std::make_unsigned_t<To>>();
+  if constexpr (has_unsigned_v<From>)
+    check_convert<std::make_unsigned_t<From>, To>();
+  if constexpr (has_unsigned_v<To> && has_unsigned_v<From>)
+    check_convert<std::make_unsigned_t<From>, std::make_unsigned_t<To>>();
+}
+
+template <typename From> void check_convert_from() {
+  check_signed_unsigned_convert_to<From, sycl::byte>();
+  // check_signed_unsigned_convert_to<From, std::byte>();
```

Comment:
I'd replicate the `FIXME` comment here, just to clarify that this is disabled on purpose.

Reply:
Replicated it in 2ba98e5bf276996a3083278ea416e4f2c2a3f89c